[feat] 북마크 좋아요 및 좋아요 해제 API 추가

ww8007 opened the pull request:
## 📌 개요 (필수)

- resolve #36 

- [x] 북마크 좋아요
- [x] 북마크 좋아요 해제

<br>

## 🔨 작업 사항 (필수)

- PUT /api/bookmarks/:bookmarkId/like
- DELETE /api/bookmarks/:bookmarkId/like

<br>

## ⚡️ 관심 리뷰 (선택)

- 특별히 확인 받고 싶은 내용을 적어주세요. 

### 1. 
bookmark entity에 다음과 같이 url 길이를 2000자 제한을 추가 했는데
괜찮은 방법일지에 대해서 궁금 합니다.
```java
  @Column(nullable = false, length = 2000)
```
### 2.

- [#30 Postgres @Lob (text type) 이슈](https://github.com/pickly-team/pickly-backend/issues/30) 내용에 대해서 처리를 했는데 확인해 주시면 감사하겠습니다 🙇‍♂️

### 3.

유저가 좋아요, 좋아요 해제 요청을 보낼 시 게시글 작성자와 API 요청자가 같은 유저인지 판별을 해야할 것 같은데
현재 Entity에 memberId가 존재하지 않아서 일단 제외 시켜 두었습니다.
토큰으로 검사하는 이외에도 API 요청에서 memberId를 추가시켜야겠죠? 

<br>

## 🌱 연관 내용 (선택)

- #30 Postgres @Lob (text type) 이슈

ww8007 committed:
fix: #30 @Lob 삭제 및 url 2000자 제한 추가
feat: 좋아요, 좋아요 해제 메서드 추가

ww8007 committed:
feat: 북마크 좋아요, 좋아요 해제 API 추가

JerryK026 commented on pickly-service/src/main/java/org/pickly/service/bookmark/controller/BookmarkController.java:
```diff
@@ -1,15 +1,49 @@
+  }
+
+
+  @DeleteMapping("/{bookmarkId}/like")
+  @Operation(summary = "Unlike bookmark", description = "Unlike bookmark")
+  public ResponseEntity<Void> unlikeBookmark(
+      @PathVariable
+      @Schema(description = "Bookmark id", example = "1")
+      Long bookmarkId
+  ) {
+    bookmarkService.unlikeBookmark(bookmarkId);
+    return ResponseEntity.ok().build();
+  }
```
unlike는 not like의 의미에 더 가까우니 cancelLike 정도로 수정해 보는 건 어떨까요?

ww8007 replied:
오 cancleLike가 더 괜찮아 보이긴 하네요 
이름 짓는게 가장 어려운 1인... 🤣

litsynp replied:
좋아요를 삭제하는 의미로 deleteLike도 말이 될 것 같아요!

EunjiShin replied:
JPA 따서 deleteLike이나 removeLike 좋을 것 같습니다!

ww8007 replied:
deleteLike로 반영 하겠습니다!

JerryK026 commented on pickly-service/src/main/java/org/pickly/service/bookmark/service/impl/BookmarkServiceImpl.java:
```diff
@@ -9,9 +11,25 @@
   private final BookmarkRepository bookmarkRepository;
 
+
+  @Override
+  @Transactional
+  public void likeBookmark(Long bookmarkId) {
+    Bookmark bookmark = bookmarkRepository.findOneById(bookmarkId)
+        .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 북마크입니다."));
+
+    bookmark.updateUserLike();
+  }
+
+  @Override
+  public void unlikeBookmark(Long bookmarkId) {
+    Bookmark bookmark = bookmarkRepository.findOneById(bookmarkId)
+        .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 북마크입니다."));
+
+    bookmark.updateUserUnlike();
+  }
```
이 부분은 얘기가 좀 필요할 것으로 보이는데 생성 / 수정한 뒤에 쓰기 작업한 대상에 대한 정보 반환을 필요하는게 좋을까요? 다들 어떻게 생각하시나요

litsynp replied:
안쓰는 데이터라면 굳이 반환하지 않아도 될 것 같긴 합니다!

EunjiShin replied:
저도 반환할 필요 없다고 생각합니다~~

ww8007 replied:
프론트에서 데이터 처리 시 변경된 내용을 가지고 북마크 포스트 전체 값을 받아서 갈아 끼우는 형식도 생각해 봤는데
굳이 좋아요에 대해서는 데이터 하나에 대해서 모든 내용을 보낼 필요는 없을 것 같아서 반환하지 않아도 괜찮겠네요!

JerryK026 reviewed (commented):
잘 봤습니다~ 굿!

Gnu-Kenny commented on pickly-service/src/main/java/org/pickly/service/bookmark/service/impl/BookmarkServiceImpl.java:
```diff
@@ -9,9 +11,25 @@
   private final BookmarkRepository bookmarkRepository;
 
+
+  @Override
+  @Transactional
+  public void likeBookmark(Long bookmarkId) {
```
좋아요 해제와 다르게 트랜잭션을 걸어두신 이유가 있을까요?

ww8007 replied:
없군요! 수정하겠습니다 😱

ww8007 replied:
아뇨! 딱히 이유는 없었네요 😱
피클리 서비스 특성상 자신의 글에만 좋아요를 누를 수 있는점에 따라서 트랜잭션 처리는 굳이 걸지 않아도 될 것 같기도 하네요
프론트 단에서 연속적으로 눌린 좋아요에 대한 처리도 추가할 수 있을 것 같고요!
좋은 지적 감사합니다 🙇‍♂️

Gnu-Kenny reviewed (commented):
코드가 전반적으로 깔끔하네요! 👍 궁금한 부분이 있어서 코멘트 하나만 남깁니다~

ww8007 commented:
추후 참고해보면 좋은 글
https://tecoble.techcourse.co.kr/post/2022-10-10-like-count/

ww8007 committed:
fix: Transactional 처리 삭제
https://github.com/pickly-team/pickly-backend/pull/37#discussion_r1121148339
refactor: unlike -> cancelLike 메서드 네이밍 변경
https://github.com/pickly-team/pickly-backend/pull/37#discussion_r1121071964

ww8007 committed:
refactor: unlike -> cancelLike 메서드 네이밍 변경
https://github.com/pickly-team/pickly-backend/pull/37#discussion_r1121071964

EunjiShin commented on pickly-service/src/main/java/org/pickly/service/bookmark/controller/BookmarkController.java:
```diff
@@ -1,15 +1,48 @@
   private final BookmarkService bookmarkService;
 
+
+  @PutMapping("/{bookmarkId}/like")
+  @Operation(summary = "Like bookmark", description = "Like bookmark")
+
+  // TODO : 멤버 아이디를 토큰에서 가져오도록 수정
+  // 또는 북마크 작업하시는 분이 멤버 아이디를 추가해 주시면 추가할 것
+  public ResponseEntity<Void> likeBookmark(
```
사소한 의견인데, annotation 위치와 주석 위치를 바꿔주면 더 좋을 것 같아요!

EunjiShin replied:
그리고 `@RestController`를 사용했기에, `ResponseEntity`로 감쌀 필요 없습니다!
성공할 때 기본 상태코드 200이 적용되고, 만약 200 말고 다른 상태 코드를 리턴하고 싶으면 `@ResponseStatus`를 이용할 수 있어요.

```
@PutMapping("/{bookmarkId}/like")
@ResponseStatus(value = HttpStatus.CREATED)
@Operation(summary = "Like bookmark", description = "Like bookmark")
public void likeBookmark(     
   @PathVariable @Schema(description = "Bookmark id", example = "1") Long bookmarkId
 ) {
    bookmarkService.likeBookmark(bookmarkId);
}
```

EunjiShin replied:
좋아요가 bookmark의 필드 update라서 Put을 사용하신거죠?
내부 동작 자체는 update가 맞긴 합니다만, '좋아요의 생성' 이라는 의미에서 Post를 써도 좋을 것 같아요.

- 좋아요 생성 : POST /bookmarks/{bookmarkId}/like
- 좋아요 해제 : DELETE /bookmarks/{bookmarkdId}/like

처럼 쓰면 HTTP method와 엔드포인트만으로 좀 더 명시적일 것 같아 의견 남겨봅니다!

ww8007 replied:
남겨주신 내용 반영했습니다!
POST, PUT 관련해서는 궁금한 점이 있는데요
좋아요를 눌러서 반영되는 API가 이전 상태(True or False)에 관계 없이 True로 항상 반영됨에 있어서
오히려 PUT이 맞지 않을까 생각이 들어서요 🧐
이에 대한 의견은 어떠세요?

EunjiShin replied:
흠 이전 상태에 관계없이 항상 같게 반환된다면, POST도 괜찮을 것 같은데요?!
PUT은 기존 리소스를 새로운 데이터로 업데이트하는 Method인데, 이 경우 기존 값이 true여도 true로 바꾸게 되니까요 🤔 좋아요의 create 라는 의미로 POST를 써도 나쁘지 않을 것 같습니다!

ww8007 replied:
저희 정책상 좋아요의 카운트가 아닌 유/무로만 판단하니 POST가 더 맞아 보이네요!
가장 크게 고민이 들었던 부분이 예전에 영한님 네트워크 강의 에서
> POST - 멱등하지 않음
메시지 바디를 통해 서버로 요청 데이터를 전달
요청 데이터를 처리, 주로 등록

> PUT - 멱등
정확한 URI를 지정하여 리소스가 있으면 대체, 생성(덮어씀)

라는 내용이 있었는데 이에 대한 개념이 정확하지 않아서 고민을 해봤습니다.
현재 코드는 POST이니 이대로 올리도록 하겠습니다 🤘

EunjiShin commented on pickly-service/src/main/java/org/pickly/service/bookmark/controller/BookmarkController.java:
```diff
@@ -1,15 +1,48 @@
+  // TODO : 멤버 아이디를 토큰에서 가져오도록 수정
+  // 또는 북마크 작업하시는 분이 멤버 아이디를 추가해 주시면 추가할 것
+  public ResponseEntity<Void> likeBookmark(
+      @PathVariable
+      @Schema(description = "Bookmark id", example = "1")
+      Long bookmarkId
```
bookmarkId는 항상 양수니까 @Positive validation을 넣어주면 좋을 것 같아요!

EunjiShin commented on pickly-service/src/main/java/org/pickly/service/bookmark/entity/Bookmark.java:
```diff
@@ -47,4 +45,11 @@ public class Bookmark extends BaseEntity {
   @Enumerated(EnumType.STRING)
   private Visibility visibility;
 
+  public void updateUserLike() {
+    this.isUserLike = true;
+  }
+
+  public void updateUserCancelLike() {
+    this.isUserLike = false;
+  }
```
- 좋아요 생성 : like
- 좋아요 해제 : removeLike 또는 unLike 또는 deleteLike

로 메소드 명을 바꿔도 좋을 것 같아요! 좋아요 추가와 해제라는 기능명을 좀 더 직관적으로 나타낼 수 있게..! 🤔

EunjiShin commented on pickly-service/src/main/java/org/pickly/service/bookmark/service/impl/BookmarkServiceImpl.java:
```diff
@@ -1,17 +1,33 @@
+
+  @Override
+  public void likeBookmark(Long bookmarkId) {
+    Bookmark bookmark = bookmarkRepository.findOneById(bookmarkId)
+        .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 북마크입니다."));
```
```
Bookmark findById(Long id) {
    return bookmarkRepository.findOneById(bookmarkId)
            .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 북마크입니다."));
  }
```

라는 별도 메소드로 분리하면 재활용성이 좋을 것 같아요!
단일 북마크 조회는 좋아요 기능 외에도 자주 사용되니까요..!

ww8007 replied:
백엔드를 처음 해보니 공통 로직을 뽑을 생각도 안해봤네요 😱
좋은 의견 감사합니다!

EunjiShin commented:
LGTM 👍👍👍
백엔드 데뷔를 축하드립니다 😎

EunjiShin commented:
1.

> bookmark entity에 다음과 같이 url 길이를 2000자 제한을 추가 했는데
괜찮은 방법일지에 대해서 궁금 합니다.

저도 다른 방법이 딱히 떠오르지않네요..🤔
엔티티엔 이렇게만 기재해두고, 이후 북마크 생성할때 request에 validate 걸어두면 될 것 같아요!


2.

> #30 Postgres @Lob (text type) 이슈 내용에 대해서 처리를 했는데 확인해 주시면 감사하겠습니다 🙇‍♂️

1번 항목 말씀하시는거죠? 넵 확인했습니다!

3.

> 유저가 좋아요, 좋아요 해제 요청을 보낼 시 게시글 작성자와 API 요청자가 같은 유저인지 판별을 해야할 것 같은데
현재 Entity에 memberId가 존재하지 않아서 일단 제외 시켜 두었습니다.
토큰으로 검사하는 이외에도 API 요청에서 memberId를 추가시켜야겠죠?

네!보통 API 접근 권한(관리자, 슈퍼유저 등)과 데이터 접근 권한 두가지를 관리합니다.
토큰 검사와 별개로 추가해주시면 될 것 같아요~

지금 구조상으론 bookmark의 category의 member를 받아와서 체크해주시면 됩니다!
(근데 제가 태스크 진행하면서 북마크 엔티티에 member FK를 고냥 추가했어요ㅋㅋ 멤버의 북마크 관련 기능이 많아서요..🙄)

일단 merge 하고, member 기능(jwt) 완성된 후 업데이트하시면 될 것 같습니다~ @ww8007

ww8007 committed:
refactor: @RestController 사용으로 ResponseEntity 삭제
@PutMapping -> @PostMapping 으로 변경,
feat: 양수임을 보장하기 위해 bookmarkId @Positive 검증 추가

ww8007 committed:
refactor: 직관적인 네이밍을 위해 like, deleteLike로 메서드 네이밍 변경

ww8007 committed:
refactor: 단일 북마크를 찾는 findById 유틸 메서드 추가

ww8007 committed:
refactor: bookmarkId @Positive message 추가
북마크 ID는 양수입니다.
